packages/preset-dumi/src/transformer/remark/code.ts

```diff
@@ -32,7 +31,42 @@ export default function code(): IDumiUnifiedTransformer {
               ...parsedAttrs,
             },
           },
-        });
+        };
+        const len = ancestor.length;
+        const grandparent = ancestor[len - 2] as IDumiElmNode;
+        const parent = ancestor[len - 1] as IDumiElmNode;
+        const nodeIndex = parent.children.indexOf(node);
+
+        // replace original node
+        parent.children.splice(nodeIndex, 1, previewer);
+
+        // https://github.com/umijs/dumi/issues/418
+        // remark-parse will create a extra paragraph ast node as container when <code src="..." /> wraps.
+        if (parent.tagName === 'p') {
+          const hoists = [] as IDumiElmNode[];
+          parent.children = parent.children.reduce((children, child, index) => {
```

Comment:
我是这个想法：

```suggestion
          grandparent.children.splice(
            grandparent.children.indexOf(parent),
            1,
            ...parent.children.reduce((result, child) => {
              if (child.properties.type === 'previewer') {
                result.push(child);
              } else {
                if (!result[result.length - 1]?.properties.type === 'previewer') {
                  result.push({
                    type: 'element',
                    tagName: 'p',
                    properties: {},
                    children: [],
                  });
                }
                
                result[result.length - 1].children.push(child);
              }
            }, [] as IDumiElmNode[]),
          );
```

这样才能确保 `<code />123<strong>456</strong>*789*` 中，`123456789` 能在同一个 `p` 里面；另外 inline elements 的判定有些太严谨了，理论上 remark 应该要顾及 DOM 嵌套的合法性，如果后面还有非 dumi 引起的 DOM 嵌套合法性问题，我们可以专门开个 remark plugin 来处理这件事

Reply:
有几个地方总结下：
1. 下面这里恒为  false（boolean === string），应该是意外写错吧？前面的 `!`会被优先计算，进而发生类型转换
     ```ts
     if (!result[result.length - 1]?.properties.type === 'previewer') {
     ```
     修改为:
     ```ts
     if (!result[result.length - 1] || result[result.length - 1].properties.type === 'previewer') {
     ```
1. 已在新的 commit 9fb52a20b2882d21a0110bd25a8e9a6529d962f6 中采用 **改编后的** https://github.com/umijs/dumi/pull/427#discussion_r543525304 方式，另外 **经测试发现**  https://github.com/umijs/dumi/pull/427#discussion_r543525304 思路有个细节需要额外处理，其 **必须** 在调整过 grandparent.children 后退出当前 parent.children 的遍历，否则最终结果会出现重复节点生成的情况，原因如 3。
1. 因为一旦发生 p 替换，那么 p.children 将替换原有 p 的位置及其相邻后续节点，那么下一次 grandparent.children 应该继续遍历 grandrent.children[pIndex+1] 的情况，结合 2，其 p 在 grandparent.children 中被替换后，p.children 并未改变，如不主动打断会继续遍历 p.children，同时提升后的 p.children 又会在 grandparent.children 后续遍历时再次遍历。
1. 结合 3，后续子集被提升，在 grandparent.children 这一层被遍历，即可以覆盖 https://github.com/umijs/dumi/pull/427#discussion_r543017830 的场景。
1. 对于后续非 dumi 的 nesting error 也是应该通过额外的 remark plugin 来集中处理比较好，所以移除了 pr 原有的行内元素判断

Reply:
本地 mac, win 均能过单测，但是 github action 环境中似乎无法正常生成 raw-demo-missing-react-6 后面的序号😭

```diff
- </div><Previewer {...require('@@/dumi/demos').default['raw-demo-missing-react-6'].previewerProps}><DumiDemo9 /></Previewer><div className="markdown">
+ </div><Previewer {...require('@@/dumi/demos').default['raw-demo-missing-react'].previewerProps}><DumiDemo9 /></Previewer><div className="markdown">
```

Reply:
👍CI 这个我看下

Reply:
看起来是重复遍历导致的？两个点可以试试看：
1. 参考 unified 的指南提升某个元素的子节点：https://unifiedjs.com/learn/recipe/remove-node/#replacing-a-node-with-its-children
2. `previewer` 节点的遍历和父级的去除分开遍历更好，两个 visit 分别处理各自的逻辑

Reply:
preview 节点创建的遍历和父级移除的遍历分开进行了遍历，其中父级移除遍历是修改为 https://github.com/umijs/dumi/pull/427#discussion_r544173946 的方式

Reply:
CI 还是挂了，看来不是这个导致的

Reply:
## 对于 ci 问题的尝试
1. wsl ubuntu 中 node v15.4 无法复现 ci 中无法生成序号的情况
2. [fork repo](https://github.com/lbwa/dumi/actions) 中无法复现 ci 中无法生成序号的情况

![image](https://user-images.githubusercontent.com/32980314/102515994-585c9a00-40c9-11eb-92b4-cdc71e2e440d.png)

Reply:
感觉是 CI 缓存导致的，我这边 Push 到另外的分支试了下也是可以的：https://github.com/umijs/dumi/runs/1575440358
要么先合并下 master 再试试看，master 有修改 previewId 的逻辑，所以某些 case 得本地看看是否要做更新

Reply:
方便的话加个钉钉呗：wuzoh5u

Reply:
本地 rebase master 后出现与 ci 相同的单测错误

Reply:
> 方便的话加个钉钉呗：wuzoh5u

已加

Reply:
> 本地 rebase master 后出现与 ci 相同的单测错误

@PeachScript  已经解决 ci 的单测问题了，是 previewer id 之前的一个修改导致的

https://github.com/umijs/dumi/blob/ec2d218e76e77f80cde413b2eda79e6a34d9981e/packages/preset-dumi/src/transformer/remark/previewer.ts#L79-L85

这里 currentIdCount  第一次取值为 undefined, 而 undefined + 1 值为 NaN，NaN + 1 仍为 NaN，进而导致 id 一直是 NaN，进而无法生成原有的 previewer id 序号。

## change

在 https://github.com/umijs/dumi/pull/427/commits/a535984afb1feec84234dd5d7347f385ea2534ad 中新增 currentIdCount 默认值，进而避免类型转换，且始终为 number 类型

**🎉现在所有 ci 测试通过👌**